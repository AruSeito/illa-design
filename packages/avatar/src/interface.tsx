import { HTMLAttributes, ReactNode } from "react"

export type AvatarColorScheme =
  string
  | "white"
  | "blackAlpha"
  | "gray"
  | "grayBlue"
  | "red"
  | "orange"
  | "yellow"
  | "green"
  | "blue"
  | "cyan"
  | "purple"

export type AvatarShape = "circle" | "square"

export type AvatarSize = "small" | "medium" | "large"

export interface AvatarProps extends HTMLAttributes<HTMLDivElement> {
  colorScheme?: AvatarColorScheme
  size?: AvatarSize
  shape?: AvatarShape
  text?: string
  src?: string
  icon?: ReactNode
}

export interface AvatarGroupProps extends Omit<AvatarProps, "text" | "src" | "icon" | "shape"> {
  zIndexAscend?: boolean
  maxCount?: number
}

export interface AvatarGroupContextProps extends Pick<AvatarGroupProps, "zIndexAscend" | "maxCount" | "colorScheme" | "size" | "style"> {
}