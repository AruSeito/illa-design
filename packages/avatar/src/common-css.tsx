import { globalColor, illaPrefix } from "@illa-design/theme"
import { AvatarProps, ColorScheme, Size } from "./interface"
import { SerializedStyles } from "@emotion/serialize"
import { css } from "@emotion/react"

export const colors: ColorScheme[] = ["white", "blackAlpha", "gray", "grayBlue", "red", "orange", "yellow", "green", "blue", "cyan", "purple"]

export function applyBgColor(colorScheme: ColorScheme): SerializedStyles {
  let bgColor: SerializedStyles
  if (colors.includes(colorScheme)) {
    if (colorScheme == "gray") {
      bgColor = css`
        background-color: ${globalColor(`--${illaPrefix}-${colorScheme}-05`)}
      `
    } else {
      bgColor = css`
        background-color: ${globalColor(`--${illaPrefix}-${colorScheme}-01`)}
      `
    }
  } else {
    bgColor = css`
      background-color: ${colorScheme};
    `
  }
  return bgColor
}

export function applyAvatarShapeSquare(size: Size): SerializedStyles {
  let width: number
  let height: number
  let textSize: number
  [width, height, textSize] = applyAvatarSize(size)
  return css`
    width: ${width}px;
    height: ${height}px;
    font-size: ${textSize}px;
    border-radius: 4px;
  `
}

export function applyAvatarShapeCircle(size: Size): SerializedStyles {
  let width: number
  let height: number
  let textSize: number
  [width, height, textSize] = applyAvatarSize(size)
  return css`
    width: ${width}px;
    height: ${height}px;
    font-size: ${textSize}px;
    border-radius: 50%;
  `
}

export function applyAvatarContainer(colorScheme: ColorScheme) {
  return css`
    ${applyBgColor(colorScheme)};
    display: flex;
    overflow: hidden;
    word-wrap: break-word;
    white-space:nowrap;
    color: ${globalColor("--illa-white-01")};
    flex-direction: row;
    align-items: center;
    justify-content: center;
    padding: 4px;
  `
}

export function applyAvatarSize(size: Size): number[] {
  let width: number
  let height: number
  let textSize: number
  switch (size) {
    case "large": {
      width = 64
      height = 64
      textSize = 24
      break
    }
    case "medium": {
      width = 40
      height = 40
      textSize = 14
      break
    }
    case "small": {
      width = 32
      height = 32
      textSize = 14
      break
    }
  }
  return [width, height, textSize]
}

export function applyMergeCss(props: AvatarProps): SerializedStyles {
  const currentColorScheme = props.colorScheme ?? "blue"
  const currentSize = props.size ?? "small"
  const currentShape = props.shape ?? "circle"

  const containerCss = applyAvatarContainer(currentColorScheme)
  const sizeCss = applyAvatarSize(currentSize)
  let shapeCss: SerializedStyles
  switch (currentShape) {
    case "circle": {
      shapeCss = applyAvatarShapeCircle(currentSize)
      break
    }
    case "square": {
      shapeCss = applyAvatarShapeSquare(currentSize)
      break
    }
  }

  return css`
    ${containerCss};
    ${sizeCss};
    ${shapeCss};
  `
}