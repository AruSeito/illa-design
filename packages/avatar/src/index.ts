export * from "./avatar"
export * from "./interface"
export * from "./avatar-group"