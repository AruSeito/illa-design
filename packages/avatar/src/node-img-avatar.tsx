/** @jsx jsx */
import { FC } from "react"
import { AvatarProps } from "./interface"
import * as React from "react"

export const NodeImgAvatar: FC<AvatarProps> = (props) => {
  return <span>
    {props.children}
  </span>
}