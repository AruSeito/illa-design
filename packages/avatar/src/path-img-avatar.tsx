/** @jsx jsx */
import { FC } from "react"
import { AvatarProps } from "./interface"
import * as React from "react"

export const PathImgAvatar: FC<AvatarProps> = (props) => {
  return <div>
    {props.text}
  </div>
}