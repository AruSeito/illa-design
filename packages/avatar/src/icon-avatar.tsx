/** @jsx jsx */
import * as React from "react"
import { FC } from "react"
import { AvatarProps, Size } from "./interface"
import { applyMergeCss } from "./common-css"
import { css } from "@emotion/react"
import { PersonIcon } from "@illa-design/icon"

function applyIconSize(size: Size) {
  let width: number
  let height: number
  switch (size) {
    case "large": {
      width = 28
      height = 28
      break
    }
    case "medium": {
      width = 18
      height = 18
      break
    }
    case "small": {
      width = 14
      height = 14
      break
    }
  }
  return css`
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    width: ${width}px;
    height: ${height}px;
  `
}

export const IconAvatar: FC<AvatarProps> = (props) => {
  const currentColorScheme = props.colorScheme ?? "blue"
  const currentSize = props.size ?? "small"
  const currentShape = props.shape ?? "circle"

  let iconSize: string
  switch (currentSize) {
    default:
    case "large": {
      iconSize = "28px"
      break
    }
    case "medium": {
      iconSize = "18px"
      break
    }
    case "small": {
      iconSize = "14px"
      break
    }
  }

  const currentIcon = props.icon ?? <PersonIcon size={iconSize} />
  const finalProps = {
    ...props,
    colorScheme: currentColorScheme,
    size: currentSize,
    shape: currentShape,
    icon: currentIcon,
  } as AvatarProps
  return <div css={applyMergeCss(finalProps)}>
      <span css={applyIconSize(finalProps.size!!)}>
        {finalProps.icon}
      </span>
  </div>
}
