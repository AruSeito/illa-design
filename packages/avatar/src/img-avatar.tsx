/** @jsx jsx */
import { FC } from "react"
import { AvatarProps, Shape } from "./interface"
import * as React from "react"
import { applyAvatarSize, applyShape } from "./common-css"
import { Image } from "@illa-design/image"

export const ImgAvatar: FC<AvatarProps> = (props) => {
  const {
    size = "small",
    shape = "circle",
  } = props

  const [width, height, _] = applyAvatarSize(size)
  return <Image src={props.src} width={width} height={height} radius={applyShape(shape)} />
}