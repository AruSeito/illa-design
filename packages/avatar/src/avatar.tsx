/** @jsxImportSource @emotion/react */
import * as React from "react"
import { forwardRef } from "react"
import { AvatarProps } from "./interface"
import { IconAvatar } from "./icon-avatar"
import { TextAvatar } from "./text-avatar"
import { ImgAvatar } from "./img-avatar"
import { omit } from "@illa-design/system"
import { AvatarGroupContext } from "./avatar-group"
import { css } from "@emotion/react"

const applyOuterCss = css`
  vertical-align: middle;
  display: inline-flex;
`

export const Avatar = forwardRef<HTMLDivElement, AvatarProps>((props, ref) => {

  const otherProps = omit(props, ["icon", "colorScheme", "size", "text", "shape", "src"])

  return <div css={applyOuterCss} ref={ref} {...otherProps}>
    <AvatarGroupContext.Consumer>
      {value => {
        let newValue = value
        if (value != undefined) {
          newValue = omit(value, ["zIndexAscend", "maxCount"])
        }
        const newProps = {
          ...newValue,
          ...props,
        }
        if (props.src != undefined) {
          return <ImgAvatar  {...newProps} />
        } else if (props.text != undefined) {
          return <TextAvatar {...newProps} />
        } else {
          return <IconAvatar {...newProps} />
        }
      }}
    </AvatarGroupContext.Consumer>
  </div>
})
