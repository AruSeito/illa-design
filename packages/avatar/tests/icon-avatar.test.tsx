import { create } from "react-test-renderer"
import { IconAvatar } from "../src/icon-avatar"
import { BsFacebook } from "react-icons/bs"

test("Icon Avatar renders correctly", () => {
  const node = create(<IconAvatar className="test-icon-avatar" size="small" icon={<BsFacebook />} />)
  expect(node.root.findByType(IconAvatar).props.className).toBe("test-icon-avatar")
})

test("Icon Avatar renders with size", () => {
  expect(create(<IconAvatar size="small" icon={<BsFacebook />} />).toJSON()).toMatchSnapshot()
  expect(create(<IconAvatar size="medium" icon={<BsFacebook />} />).toJSON()).toMatchSnapshot()
  expect(create(<IconAvatar size="large" icon={<BsFacebook />} />).toJSON()).toMatchSnapshot()
})