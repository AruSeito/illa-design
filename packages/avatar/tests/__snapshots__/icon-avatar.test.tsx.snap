// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon Avatar renders with size 1`] = `
.emotion-0 {
  background-color: #2a3ed6;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  overflow: hidden;
  word-wrap: break-word;
  white-space: nowrap;
  color: #fff;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding: 4px;
  width: 32px;
  height: 32px;
  font-size: 14px;
  line-height: 22px;
  border-radius: 50%;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 14px;
  height: 14px;
}

<div
  className="emotion-0"
>
  <div
    className="emotion-1"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16 8.049c0-4.446-3.582-8.05-8-8.05C3.58 0-.002 3.603-.002 8.05c0 4.017 2.926 7.347 6.75 7.951v-5.625h-2.03V8.05H6.75V6.275c0-2.017 1.195-3.131 3.022-3.131.876 0 1.791.157 1.791.157v1.98h-1.009c-.993 0-1.303.621-1.303 1.258v1.51h2.218l-.354 2.326H9.25V16c3.824-.604 6.75-3.934 6.75-7.951z"
      />
    </svg>
  </div>
</div>
`;

exports[`Icon Avatar renders with size 2`] = `
.emotion-0 {
  background-color: #2a3ed6;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  overflow: hidden;
  word-wrap: break-word;
  white-space: nowrap;
  color: #fff;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding: 4px;
  width: 40px;
  height: 40px;
  font-size: 14px;
  line-height: 22px;
  border-radius: 50%;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 18px;
  height: 18px;
}

<div
  className="emotion-0"
>
  <div
    className="emotion-1"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16 8.049c0-4.446-3.582-8.05-8-8.05C3.58 0-.002 3.603-.002 8.05c0 4.017 2.926 7.347 6.75 7.951v-5.625h-2.03V8.05H6.75V6.275c0-2.017 1.195-3.131 3.022-3.131.876 0 1.791.157 1.791.157v1.98h-1.009c-.993 0-1.303.621-1.303 1.258v1.51h2.218l-.354 2.326H9.25V16c3.824-.604 6.75-3.934 6.75-7.951z"
      />
    </svg>
  </div>
</div>
`;

exports[`Icon Avatar renders with size 3`] = `
.emotion-0 {
  background-color: #2a3ed6;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  overflow: hidden;
  word-wrap: break-word;
  white-space: nowrap;
  color: #fff;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding: 4px;
  width: 64px;
  height: 64px;
  font-size: 24px;
  line-height: 32px;
  border-radius: 50%;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 28px;
  height: 28px;
}

<div
  className="emotion-0"
>
  <div
    className="emotion-1"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16 8.049c0-4.446-3.582-8.05-8-8.05C3.58 0-.002 3.603-.002 8.05c0 4.017 2.926 7.347 6.75 7.951v-5.625h-2.03V8.05H6.75V6.275c0-2.017 1.195-3.131 3.022-3.131.876 0 1.791.157 1.791.157v1.98h-1.009c-.993 0-1.303.621-1.303 1.258v1.51h2.218l-.354 2.326H9.25V16c3.824-.604 6.75-3.934 6.75-7.951z"
      />
    </svg>
  </div>
</div>
`;
